x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/factories/utils/transform_hit_to_alert.ts

```diff
@@ -0,0 +1,143 @@
+  doc: estypes.SearchHit<SignalSource>;
+  mergeStrategy: ConfigType['alertMergeStrategy'];
+  ignoreFields: Record<string, boolean>;
+  ignoreFieldsRegexes: string[];
+  applyOverrides: boolean;
+  buildReasonMessage: BuildReasonMessage;
+  indicesToQuery: string[];
+  alertTimestampOverride: Date | undefined;
+  ruleExecutionLogger: IRuleExecutionLogForExecutors;
+  alertUuid: string;
+  publicBaseUrl?: string;
+}
+
+/**
+ * Formats the search_after result for insertion into the signals index. We first create a
+ * "best effort" merged "fields" with the "_source" object, then build the signal object,
+ * then the event object, and finally we strip away any additional temporary data that was added
+ * such as the "threshold_result".
+ * @param completeRule The rule saved object to build overrides
+ * @param doc The SignalSourceHit with "_source", "fields", and additional data such as "threshold_result"
+ * @returns The body that can be added to a bulk call for inserting the signal.
+ */
+export const transformHitToAlert = ({
+  spaceId,
+  completeRule,
+  doc,
+  mergeStrategy,
+  ignoreFields,
+  ignoreFieldsRegexes,
+  applyOverrides,
+  buildReasonMessage,
+  indicesToQuery,
+  alertTimestampOverride,
+  ruleExecutionLogger,
+  alertUuid,
+  publicBaseUrl,
+}: TransformHitToAlertProps): BaseFieldsLatest => {
+  const mergedDoc = getMergeStrategy(mergeStrategy)({ doc, ignoreFields, ignoreFieldsRegexes });
+  const thresholdResult = mergedDoc._source?.threshold_result;
+
+  if (isSourceDoc(mergedDoc)) {
+    const overrides = applyOverrides
+      ? {
+          nameOverride: buildRuleNameFromMapping({
+            eventSource: mergedDoc._source ?? {},
+            ruleName: completeRule.ruleConfig.name,
+            ruleNameMapping: completeRule.ruleParams.ruleNameOverride,
+          }).ruleName,
+          severityOverride: buildSeverityFromMapping({
+            eventSource: mergedDoc._source ?? {},
+            severity: completeRule.ruleParams.severity,
+            severityMapping: completeRule.ruleParams.severityMapping,
+          }).severity,
+          riskScoreOverride: buildRiskScoreFromMapping({
+            eventSource: mergedDoc._source ?? {},
+            riskScore: completeRule.ruleParams.riskScore,
+            riskScoreMapping: requiredOptional(completeRule.ruleParams.riskScoreMapping),
+          }).riskScore,
+        }
+      : undefined;
+
+    const reason = buildReasonMessage({
+      name: overrides?.nameOverride ?? completeRule.ruleConfig.name,
+      severity: overrides?.severityOverride ?? completeRule.ruleParams.severity,
+      mergedDoc,
+    });
+
+    const alertFields = buildAlertFields({
+      docs: [mergedDoc],
+      completeRule,
+      spaceId,
+      reason,
+      indicesToQuery,
+      alertUuid,
+      publicBaseUrl,
+      alertTimestampOverride,
+      overrides,
+    });
+
+    const {
+      result: validatedSource,
+      removed: removedSourceFields,
+      fieldsToAdd,
```

Comment:
We could, but to insert the new fields in an arbitrary location `traverseDoc` would need to also keep a reference to the top level object as it descends through the levels recursively. I felt like that would make the recursive part harder to follow since right now the only "global context" the recursive function needs is the `path` - the mutations on each sub-object are independent from any parents.

Reply:
it can be implemented within 

```
export const traverseAndMutateDoc = <T extends SourceFieldRecord>(document: T) => {
  return internalTraverseAndMutateDoc({
    document,
    path: [],
    topLevel: true,
    removed: [],
    fieldsToAdd: [],
  });
};
```

something like this, encapsulating this logic within `traverseAndMutateDoc` and `internalTraverseAndMutateDoc` would be the same.


```ts
export const traverseAndMutateDoc = <T extends SourceFieldRecord>(document: T) => {
  const traversedDoc = internalTraverseAndMutateDoc({
    document,
    path: [],
    topLevel: true,
    removed: [],
    fieldsToAdd: [],
  });

    traversedDoc.fieldsToAdd.forEach(({ key, value }) => {
      traversedDoc. result[key] = value;
    });

 return traversedDoc;
};

```

Reply:
Ah I see, yeah I like that better. Made the change 👍